File: functions/utils/logging_utils.py
import logging

def get_logger(name):
    """
    Creates and returns a logger with the specified name.
    
    This utility function provides a standardized way to create loggers
    across the application, ensuring consistent formatting and behavior.
    
    Args:
        name: The name for the logger, typically __name__ from the calling module
        
    Returns:
        A configured logger instance
    """
    logger = logging.getLogger(name)
    
    # Only configure handlers if they haven't been added yet
    if not logger.handlers:
        # Set the logging level
        logger.setLevel(logging.INFO)
        
        # Create a handler for console output
        handler = logging.StreamHandler()
        
        # Create a formatter and set it for the handler
        formatter = logging.Formatter('%(asctime)s - %(name)s - %(levelname)s - %(message)s')
        handler.setFormatter(formatter)
        
        # Add the handler to the logger
        logger.addHandler(handler)
    
    return logger

File: functions/src/utils/logging-utils.ts
/**
 * Log levels in order of severity
 */
const LogLevel = {
  DEBUG: 0,
  INFO: 1,
  WARN: 2,
  ERROR: 3,
} as const;

// Set minimum log level here
const MIN_LOG_LEVEL = LogLevel.WARN;

/**
 * Creates and returns a logger with the specified name.
 *
 * This utility function provides a standardized way to create loggers
 * across the application, ensuring consistent formatting and behavior.
 *
 * @param name - The name for the logger, typically __filename from the calling module
 * @returns A configured logger instance with consistent formatting
 */
export const getLogger = (name: string) => {
  // Format the name to be more readable (remove file extension and path)
  const formattedName = name.split('/').pop()?.replace('.ts', '') || name;

  return {
    info: (message: string, ...args: any[]) => {
      if (MIN_LOG_LEVEL <= LogLevel.INFO) {
        const timestamp = new Date().toISOString();
        console.log(
          `[${timestamp}] [${formattedName}] [INFO] ${message}`,
          ...args,
        );
      }
    },
    warn: (message: string, ...args: any[]) => {
      if (MIN_LOG_LEVEL <= LogLevel.WARN) {
        const timestamp = new Date().toISOString();
        console.warn(
          `[${timestamp}] [${formattedName}] [WARN] ${message}`,
          ...args,
        );
      }
    },
    error: (message: string, ...args: any[]) => {
      if (MIN_LOG_LEVEL <= LogLevel.ERROR) {
        const timestamp = new Date().toISOString();
        console.error(
          `[${timestamp}] [${formattedName}] [ERROR] ${message}`,
          ...args,
        );
      }
    },
    debug: (message: string, ...args: any[]) => {
      if (MIN_LOG_LEVEL <= LogLevel.DEBUG) {
        const timestamp = new Date().toISOString();
        console.debug(
          `[${timestamp}] [${formattedName}] [DEBUG] ${message}`,
          ...args,
        );
      }
    },
  };
};
